pyiron_base/database/filetable.py

```diff
@@ -47,11 +47,38 @@
 }
 
 
-class FileTable(IsDatabase, metaclass=Singleton):
-    def __init__(self, project):
+class FileTableSingleton(ABCMeta):
+    """
+    Indexing the file system for each `FileTable` can be expensive, so we use a
+    singleton system that does this once for each path instead.
+    """
+
+    _instances = {}
+
+    def __call__(cls, index_from):
+        _path = os.path.abspath(index_from)
+        if _path not in cls._instances:
+            cls._instances[_path] = super(FileTableSingleton, cls).__call__(index_from)
+        return cls._instances[_path]
```

Comment:
I like this implementation, there is just one case where it is inefficient: 
```
pr_a = Project("a")
pr_b = Project("a/b")
```
We can imagine the sub project `a/b` containing all the data, while `a` contains just three groups `a/b`, `a/c` and `a/d`. In this case we waste resources as the whole index has to be calculated twice. But maybe that is something that can be more easily fixed from the `pyfileindex` level.

Reply:
I think we will need at least some edits here, right, because the `PyFileIndex` objects instantiated by `FileTable` don't know about each other? I can imagine that if the `_path` is a subset of and `_instances.keys()` path, that we extract the `PyFileIndex` attribute from the super-path instance, slice it (or otherwise reduce it down, I'm not exactly sure the details of `PyFileIndex`) and use _that_ (or a copy of it) when instantiating the new `FileTable` instead of creating the new `PyFileIndex` object from scratch.

Reply:
> I think we will need at least some edits here, right, because the `PyFileIndex` objects instantiated by `FileTable` don't know about each other? 

Yes. 

> I can imagine that if the `_path` is a subset of and `_instances.keys()` path, that we extract the `PyFileIndex` attribute from the super-path instance, slice it (or otherwise reduce it down, I'm not exactly sure the details of `PyFileIndex`) and use _that_ (or a copy of it) when instantiating the new `FileTable` instead of creating the new `PyFileIndex` object from scratch.

I started working on this in https://github.com/pyiron/pyfileindex/pull/74/

Reply:
For the case that we have multiple overlapping paths we could use: 
```
max([p for p in cls._instances.keys() if os.path.commonpath([_path, p]) == p], key=len)
```

Still this needs some additional modifications for the FileTable to accept an existing `PyFileIndex`.

Reply:
Yes. I would suggest to merge this as-is, so that we can instantiate mutliple projects without failure, then open a second PR with changes to `FileTable` to allow instantiating from an existing `PyFileIndex` object.